### PR TITLE
Eye Module by Ray

### PR DESCRIPTION
**Pull request by bradymiller**
This thing is almost 2 years in the making by Ray and is way beyond impressive.

Can use it be selecting Eye Form within the encounter.

Still some minimal integration work to do (goal is to get this into the codebase very soon for the 5.0.0 release).

Demo is here:
http://www.open-emr.org/wiki/index.php/Development_Demo#192.168.1.132

@ophthal

### CODE REVIEW

**Comment by bradymiller** [comment]
here's the online demo for this:
http://www.open-emr.org/wiki/index.php/Development_Demo#192.168.1.132

File: interface/forms/eye_mag/php/taskman_functions.php [hunk 1]
```diff
@@ -0,0 +1,467 @@
...
+	$doc_id 	= $ajax_req['doc_id'];
+	$enc 	 	= $ajax_req['enc'];
+
+	//what if the document was deleted, cause we want to redo it...
+	//this checks form_taskman.  Maybe we need to check Documents...
+	$sql = "SELECT * from form_taskman where FROM_ID=? and TO_ID=? and PATIENT_ID=? and ENC_ID=?";
+	$task = sqlQuery($sql,array($from_id,$to_id,$patient_id,$enc));
+
+	if ($task['ID'] && $task['COMPLETED'] =='2') {
+		$send['comments'] = xlt('This fax has already been sent.  Consider printing the Fax Report and sending it manually.');
```
Comment (bradymiller) [inline interface/forms/eye_mag/php/taskman_functions.php]: TODO for me:
remove the double space in this translation

File: interface/forms/eye_mag/php/taskman_functions.php [hunk 2]
```diff
@@ -0,0 +1,467 @@
...
+							";
+							//add a resend option.
+			$send['comments'] = xlt('This fax has already been sent.');
+			echo json_encode($send);
+			exit;
+		} else if ($task['doc_type'] == "Fax-resend") {
+			//we need to resend this fax????
+			//You can only resend from here once.
+			//If you are that messed up, print the report and fax it manually.
+			$send['comments'] = xlt('OK, we want to resend it.  We just do not know how to do this yet automatically, so you have do it manually...');
```
Comment (bradymiller) [inline interface/forms/eye_mag/php/taskman_functions.php]: TODO for me:
remove the double space in this translation

File: interface/forms/eye_mag/php/taskman_functions.php [hunk 3]
```diff
@@ -0,0 +1,467 @@
...
+	update_taskman($task,'created', '1');
+	if ($task['DOC_TYPE'] == 'Fax') {
+		deliver_document($task);
+	}	
+	update_taskman($task,'completed', '1');
+	
+	if ($task['DOC_TYPE'] == "Fax") {
+		//now return any objects you need to Eye Form
+		$send['DOC_link'] = "<a href='".$webroot."/openemr/controller.php?document&view&patient_id=".$task['PATIENT_ID']."&doc_id=".$task['DOC_ID']."'
+								target='_blank' title=".xlt('Report was faxed.  Click to view.').">
```
Comment (bradymiller) [inline interface/forms/eye_mag/php/taskman_functions.php]: TODO for me:
remove the double space in this translation

File: interface/forms/eye_mag/php/taskman_functions.php [hunk 4]
```diff
@@ -0,0 +1,467 @@
...
+		$query = "select id from categories where name =?";
+    	$ID = sqlQuery($query,array($category_name));
+    	$category_id = $ID['id'];
+		
+		$filename = "Report_".$encounter.".pdf"; 
+		foreach (glob($filepath.'/'.$filename) as $file) {
+    	  	unlink($file); //maybe shorten to just unlink($filepath.'/'.$filename); - well this does test to see if it is there
+    	}
+    	
+    	$sql = "DELETE from categories_to_documents where document_id IN (SELECT id from documents where documents.url like '%".$filename."')";
```
Comment (bradymiller) [inline interface/forms/eye_mag/php/taskman_functions.php]: TODO for me:
sql binding

File: interface/patient_file/report/custom_report.php [hunk 5]
```diff
@@ -149,11 +149,11 @@ function postToGet($arin) {
     border-bottom: solid thin #6D6D6D;
     padding:0% 2% 0% 2.5%;
   }
+  img { max-width:700px; }
 </style>
 
 <?php if (!$PDF_OUTPUT) { ?>
 
-<script type="text/javascript" src="<?php echo $GLOBALS['assets_static_relative']; ?>/jquery-min-1-5/index.js"></script>
```
Comment (bradymiller) [inline interface/patient_file/report/custom_report.php]: TODO for me:
Add this back if needed and test this script.

File: interface/usergroup/addrbook_edit.php [hunk 6]
```diff
@@ -29,6 +29,7 @@
  $info_msg = "";
 
  function invalue($name) {
+  if (!$_POST[$name]) return "''";
```
Comment (bradymiller) [inline interface/usergroup/addrbook_edit.php]: @ophthal
What this to fix a bug?

File: library/restoreSession.php [hunk 7]
```diff
@@ -6,7 +6,7 @@
 //
 var oemr_session_name = '<?php echo session_name(); ?>';
 var oemr_session_id   = '<?php echo session_id(); ?>';
-var oemr_dialog_close_msg = '<?php echo xl('OK to close this other popup window?'); ?>';
+var oemr_dialog_close_msg = '<?php echo "OK to close this other popup window?"; ?>';
```
Comment (bradymiller) [inline library/restoreSession.php]: TODO for me:
Use the xla() function if it exists via function_exists('xla') call

File: sql/4_2_2-to-5_0_0_upgrade.sql [hunk 8]
```diff
@@ -807,3 +807,1311 @@ INSERT INTO `list_options` (`list_id`, `option_id`, `title`, `seq`, `notes`, `ac
...
+`CTLSUPPLIEROD` varchar(25) DEFAULT NULL,
+`CTLSUPPLIEROS` varchar(25) DEFAULT NULL,
+`CTLBRANDOD` varchar(50) DEFAULT NULL,
+`CTLBRANDOS` varchar(50) DEFAULT NULL,
+`ODDIAM` varchar(50) DEFAULT NULL,
+`ODBC` varchar(50) DEFAULT NULL,
+`OSDIAM` varchar(50) DEFAULT NULL,
+`OSBC` varchar(50) DEFAULT NULL,
+`RXCOMMENTS` text,
+`COMMENTS` text DEFAULT NULL,
```
Comment (bradymiller) [inline sql/4_2_2-to-5_0_0_upgrade.sql]: TODO for me:
change this to a bare TEXT

File: sql/4_2_2-to-5_0_0_upgrade.sql [hunk 9]
```diff
@@ -807,3 +807,1311 @@ INSERT INTO `list_options` (`list_id`, `option_id`, `title`, `seq`, `notes`, `ac
...
+  `ODMPDD` varchar(20) DEFAULT NULL,
+  `ODMPDN` varchar(20) DEFAULT NULL,
+  `OSMPDD` varchar(20) DEFAULT NULL,
+  `OSMPDN` varchar(20) DEFAULT NULL,
+  `BPDD` varchar(20) DEFAULT NULL,
+  `BPDN` varchar(20) DEFAULT NULL,
+  `LENS_MATERIAL` varchar(20) DEFAULT NULL,
+  `LENS_TREATMENTS` varchar(100) DEFAULT NULL,
+  `RX_TYPE` varchar(25) DEFAULT NULL,
+  `COMMENTS` text DEFAULT NULL,
```
Comment (bradymiller) [inline sql/4_2_2-to-5_0_0_upgrade.sql]: TODO for me:
make this a bare TEXT

File: interface/usergroup/addrbook_edit.php [hunk 10]
```diff
@@ -29,6 +29,7 @@
...
   $fld = add_escape_custom(trim($_POST[$name]));
   return "'$fld'";
```
Comment (bradymiller) [inline interface/usergroup/addrbook_edit.php]: @ophthal
Is the suffix stuff needed for the eye form? (I am not seeing this in the database)

File: sql/database.sql [hunk 11]
```diff
@@ -2891,7 +2903,7 @@ CREATE TABLE `list_options` (
   `toggle_setting_1` tinyint(1) NOT NULL default '0',
   `toggle_setting_2` tinyint(1) NOT NULL default '0',
   `activity` TINYINT DEFAULT 1 NOT NULL,
-  `subtype` varchar(31) NOT NULL DEFAULT '',
+  `subtype` varchar(10) DEFAULT NULL,
```
Comment (bradymiller) [inline sql/database.sql]: TODO for me:
revert this change

File: sql/database.sql [hunk 12]
```diff
@@ -245,6 +244,19 @@ INSERT INTO `categories` VALUES (10, 'Patient Photograph', '', 4, 8, 9);
...
+INSERT INTO `categories` VALUES (17, 'Imaging - Eye', '', 14, 30, 49);
+INSERT INTO `categories` VALUES (18, 'OCT - Eye', 'POSTSEG', 17, 31, 32);
+INSERT INTO `categories` VALUES (19, 'FA/ICG - Eye', 'POSTSEG', 17, 33, 34);
+INSERT INTO `categories` VALUES (20, 'External Photos - Eye', 'EXT', 17, 35, 36);
+INSERT INTO `categories` VALUES (21, 'AntSeg Photos - Eye', 'ANTSEG', 17, 37, 38);
+INSERT INTO `categories` VALUES (22, 'Optic Disc - Eye', 'POSTSEG', 17, 39, 40);
+INSERT INTO `categories` VALUES (23, 'Fundus - Eye', 'POSTSEG', 17, 41, 42);
+INSERT INTO `categories` VALUES (24, 'Radiology - Eye', 'NEURO', 17, 43, 44);
+INSERT INTO `categories` VALUES (25, 'VF - Eye', 'NEURO', 17, 45, 46);
+INSERT INTO `categories` VALUES (26, 'Drawings - Eye', '', 17, 47, 48);
```
Comment (bradymiller) [inline sql/database.sql]: TODO for me:
update the categories_seq to be 26

File: sql/database.sql [hunk 13]
```diff
@@ -6423,6 +6455,7 @@ CREATE TABLE `users` (
   `fname` varchar(255) default NULL,
   `mname` varchar(255) default NULL,
   `lname` varchar(255) default NULL,
+  `suffix` varchar(255) default NULL,
```
Comment (bradymiller) [inline sql/database.sql]: @ophthal
This is not in the upgrade script. Is this needed for the eye module (if not, then will postpone this)?

File: sql/database.sql [hunk 14]
```diff
@@ -8345,3 +8982,591 @@ CREATE TABLE calendar_external (
...
+`CTLSUPPLIEROD` varchar(25) DEFAULT NULL,
+`CTLSUPPLIEROS` varchar(25) DEFAULT NULL,
+`CTLBRANDOD` varchar(50) DEFAULT NULL,
+`CTLBRANDOS` varchar(50) DEFAULT NULL,
+`ODDIAM` varchar(50) DEFAULT NULL,
+`ODBC` varchar(50) DEFAULT NULL,
+`OSDIAM` varchar(50) DEFAULT NULL,
+`OSBC` varchar(50) DEFAULT NULL,
+`RXCOMMENTS` text,
+`COMMENTS` text DEFAULT NULL,
```
Comment (bradymiller) [inline sql/database.sql]: TODO for me:
make TEXT bare

File: sql/database.sql [hunk 15]
```diff
@@ -8345,3 +8982,591 @@ CREATE TABLE calendar_external (
...
+  `ODMPDD` varchar(20) DEFAULT NULL,
+  `ODMPDN` varchar(20) DEFAULT NULL,
+  `OSMPDD` varchar(20) DEFAULT NULL,
+  `OSMPDN` varchar(20) DEFAULT NULL,
+  `BPDD` varchar(20) DEFAULT NULL,
+  `BPDN` varchar(20) DEFAULT NULL,
+  `LENS_MATERIAL` varchar(20) DEFAULT NULL,
+  `LENS_TREATMENTS` varchar(100) DEFAULT NULL,
+  `RX_TYPE` varchar(25) DEFAULT NULL,
+  `COMMENTS` text DEFAULT NULL,
```
Comment (bradymiller) [inline sql/database.sql]: TODO for me:
make TEXT bare

File: sql/4_2_2-to-5_0_0_upgrade.sql [hunk 16]
```diff
@@ -807,3 +807,1311 @@ INSERT INTO `list_options` (`list_id`, `option_id`, `title`, `seq`, `notes`, `ac
 INSERT INTO `list_options` (`list_id`, `option_id`, `title`, `seq`, `notes`, `activity`) VALUES ('page_validation', 'facilities_add#facility-add', '/interface/usergroup/facilities_add.php', 100, '{facility:{presence: true}, ncolor:{presence: true}}', 1);
 #EndIf
```
Comment (bradymiller) [inline sql/4_2_2-to-5_0_0_upgrade.sql]: TODO for me:
extensive testing of upgrade script

File: sql/database.sql (hunk 13 above)
Comment (bradymiller) [inline sql/database.sql]: Actually, I noted it is used in your code so will bring it in also via the upgrade script.

TODO for me:
Add this to the sql upgrade script.

File: sql/4_2_2-to-5_0_0_upgrade.sql [hunk 17]
```diff
@@ -807,3 +807,1311 @@ INSERT INTO `list_options` (`list_id`, `option_id`, `title`, `seq`, `notes`, `ac
...
```
Comment (bradymiller) [inline sql/4_2_2-to-5_0_0_upgrade.sql]: TODO for me:
install the registry for the eye form on the upgrade script.

File: interface/usergroup/addrbook_edit.php (hunk 10 above)
Comment (bradymiller) [inline interface/usergroup/addrbook_edit.php]: Disregard above question. Noted it is used your form.

Comment (ophthal) [inline interface/usergroup/addrbook_edit.php]: I have an update to add suffix to the address book.
Letters look funny without the suffix and the address book looks funny with fname: John lmane: Smith, MD.  The suffix needs a home...

**Comment by bradymiller** [comment]
Only 3 issues left before I bring this into the codebase:
1. Test the upgrade to ensure the registry add of the eye form works
2. Separate out code in interface/forms/newpatient/common.php to another PR (assuming it is not necessary for the eye module)
3. test to ensure the suffix stuff is working on the address book

ETA to get this into the main codebase is tomorrow :)

File: interface/usergroup/addrbook_edit.php (hunk 6 above)
Comment (bradymiller) [inline interface/usergroup/addrbook_edit.php]: Will bring this in.
DONE

File: interface/usergroup/addrbook_edit.php (hunk 10 above)
Comment (bradymiller) [inline interface/usergroup/addrbook_edit.php]: brought this in and it tested well on the address book.
DONE

**Comment by bradymiller** [comment]
This work was committed to the main codebase. Thank you Ray for an awesome contribution.

Rebased commit here:
https://github.com/openemr/openemr/commit/00aa8a6f2cd12e668532c313fa028a56144bf62e

Followed by a commit with improvements by Ray here:
https://github.com/openemr/openemr/commit/81f10acabaf63394a1ffbf1fe837884457527488
